gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceExtractor.java

```diff
@@ -979,39 +1118,33 @@ private void fetchResultBatchWithRetry(RecordSetList<JsonElement> rs)
   @Override
   public void closeConnection() throws Exception {
     if (this.bulkConnection != null
-        && !this.bulkConnection.getJobStatus(this.bulkJob.getId()).getState().toString().equals("Closed")) {
+        && !this.bulkConnection.getJobStatus(this.getBulkJobId()).getState().toString().equals("Closed")) {
       log.info("Closing salesforce bulk job connection");
-      this.bulkConnection.closeJob(this.bulkJob.getId());
+      this.bulkConnection.closeJob(this.getBulkJobId());
     }
   }
 
-  public static List<Command> constructGetCommand(String restQuery) {
+  private static List<Command> constructGetCommand(String restQuery) {
     return Arrays.asList(new RestApiCommand().build(Arrays.asList(restQuery), RestApiCommandType.GET));
   }
 
   /**
    * Waits for the PK batches to complete. The wait will stop after all batches are complete or on the first failed batch
    * @param batchInfoList list of batch info
-   * @param retryInterval the polling interval
+   * @param waitInterval the polling interval
    * @return the last {@link BatchInfo} processed
    * @throws InterruptedException
    * @throws AsyncApiException
    */
-  private BatchInfo waitForPkBatches(BatchInfoList batchInfoList, int retryInterval)
+  private BatchInfo waitForPkBatches(BatchInfoList batchInfoList, int waitInterval)
       throws InterruptedException, AsyncApiException {
     BatchInfo batchInfo = null;
     BatchInfo[] batchInfos = batchInfoList.getBatchInfo();
 
-    // Wait for all batches other than the first one. The first one is not processed in PK chunking mode
-    for (int i = 1; i < batchInfos.length; i++) {
-      BatchInfo bi = batchInfos[i];
-
-      // get refreshed job status
-      bi = this.bulkConnection.getBatchInfo(this.bulkJob.getId(), bi.getId());
-
-      while ((bi.getState() != BatchStateEnum.Completed)
-          && (bi.getState() != BatchStateEnum.Failed)) {
-        Thread.sleep(retryInterval * 1000);
+    for (BatchInfo bi: batchInfos) {
```

Comment:
Is it no longer true that the first BatchInfo is for the parent and not the pk chunking batches or are you considering it safe to apply the same check since the state should be completed?

Reply:
I had a ticket for this - https://jira01.corp.linkedin.com:8443/browse/DSS-22221
pkchunking is using this function.
Sometimes, the parent may not be the first element in the list. (see screenshot in ticket)